src/ripple/app/misc/Validations.cpp

```diff
@@ -429,25 +429,34 @@ class ValidationsImp : public Validations
         bool anyNew = false;
 
         JLOG (j_.info()) << "Flushing validations";
-        ScopedLockType sl (mLock);
-        for (auto& it: mCurrentValidations)
         {
-            if (it.second)
-                mStaleValidations.push_back (it.second);
-
-            anyNew = true;
-        }
-        mCurrentValidations.clear ();
+            ScopedLockType sl (mLock);
+            for (auto& it: mCurrentValidations)
+            {
+                if (it.second)
+                {
+                    mStaleValidations.push_back (it.second);
+                    anyNew = true;
+                }
+            }
+            mCurrentValidations.clear ();
 
-        if (anyNew)
-            condWrite ();
+            // If there isn't a write in progress already, then write to the
+            // database synchronously.
+            if (anyNew && !mWriting)
+            {
+                mWriting = true;
+                doWrite (sl);
+            }
 
-        while (mWriting)
-        {
-            ScopedUnlockType sul (mLock);
-            std::this_thread::sleep_for (std::chrono::milliseconds (100));
+            // Handle the case where flush() is called while a queuedWrite
+            // is already in progress.
+            while (mWriting)
+            {
+                ScopedUnlockType sul (mLock);
+                std::this_thread::sleep_for (std::chrono::milliseconds (100));
+            }
```

Comment:
My understanding of the standard is that any unsynchronized access (read or write) of a non-atomic by two different threads is undefined behavior.  If my understanding is correct, then since `mWriting` is a standard `bool` it needs the lock.  Undefined behavior is nobody's friend (despite what the optimizer folks will tell you).

However I could convert `mWriting` into an `std::atomic<bool>`and we could release the lock after the call to `doWrite`.  I'm not convinced this would be a big win, since `flush()` only occurs once during the life of the program.  Making `mWriting` an `atomic` has the disadvantage that it constrains the optimizer in other parts of the code.  But it could certainly be done.  Shall I convert `mWriting` to an `atomic`?